tensorboard/defs/web.bzl

```diff
@@ -14,259 +14,295 @@
+                "tf_ts_library instead.",
+            )
+        else:
+            web_srcs.append(src)
 
-  # perform strict dependency checking
-  manifest = _make_manifest(ctx, manifest_srcs)
-  webpaths = depset(new_webpaths, transitive=[webpaths])
-  dummy, manifests = _run_webfiles_validator(ctx, web_srcs, deps, manifest)
-  web_srcs.append(dummy)
+    # perform strict dependency checking
+    manifest = _make_manifest(ctx, manifest_srcs)
+    webpaths = depset(new_webpaths, transitive = [webpaths])
+    dummy, manifests = _run_webfiles_validator(ctx, web_srcs, deps, manifest)
+    web_srcs.append(dummy)
 
-  # define development web server that only applies to this transitive closure
-  if ctx.attr.srcs:
-    devserver_manifests = manifests
-    export_deps = []
-  else:
-    # If a rule exists purely to export other build rules, then it's
-    # appropriate for the exported sources to be included in the
-    # development web server.
-    export_deps = unfurl(ctx.attr.exports)
-    devserver_manifests = depset(
-        order = "postorder",
-        transitive = (
-            [manifests] + [dep.webfiles.manifests for dep in export_deps]
+    # define development web server that only applies to this transitive closure
+    if ctx.attr.srcs:
+        devserver_manifests = manifests
+        export_deps = []
+    else:
+        # If a rule exists purely to export other build rules, then it's
+        # appropriate for the exported sources to be included in the
+        # development web server.
+        export_deps = unfurl(extract_providers(ctx.attr.exports))
+        devserver_manifests = depset(
+            order = "postorder",
+            transitive = (
+                [manifests] + [dep[WebFilesInfo].manifests for dep in export_deps]
+            ),
+        )
+    params = struct(
+        label = str(ctx.label),
+        bind = "localhost:6006",
+        manifest = [long_path(ctx, man) for man in devserver_manifests.to_list()],
+        external_asset = [
+            struct(webpath = k, path = v)
+            for k, v in ctx.attr.external_assets.items()
+        ],
+    )
+    params_file = _new_file(ctx, "-params.pbtxt")
+    ctx.actions.write(output = params_file, content = params.to_proto())
+    ctx.actions.write(
+        is_executable = True,
+        output = ctx.outputs.executable,
+        content = "#!/bin/sh\nexec %s %s" % (
+            ctx.executable._WebfilesServer.short_path,
+            long_path(ctx, params_file),
+        ),
+    )
+    params = struct(
+        label = str(ctx.label),
+        bind = "localhost:6006",
+        manifest = [long_path(ctx, man) for man in devserver_manifests.to_list()],
+        external_asset = [
+            struct(webpath = k, path = v)
+            for k, v in ctx.attr.external_assets.items()
+        ],
+    )
+    params_file = _new_file(ctx, "-params.pbtxt")
+    ctx.actions.write(output = params_file, content = proto.encode_text(params))
```

Comment:
This block (lines 123-141) seems to be repeated right above (lines 104-122).

AFAICT, the only difference is the way this method is called: `proto.encode_text(params)` vs `params.to_proto()`... probably a bad sync?

Reply:
Fixed.